lib/BuildSystem/SwiftTools.cpp

```diff
@@ -624,8 +630,322 @@ class SwiftCompilerTool : public Tool {
+      {
+        //get pid
+        if (!nodeIsScalarString(it->getKey(), "pid")) {
+          error("expected key 'pid'");
+          return false;
+        }
+        if (it->getValue()->getType() != llvm::yaml::Node::NK_Scalar) {
+          error("unexpected 'pid' value (expected scalar)");
+          return false;
+        }
+        int64_t pid;
+        if(!parseInteger(static_cast<llvm::yaml::ScalarNode*>(it->getValue()), &pid)) {
+          return false;
+        }
+        message_out.setPid(pid);
+        ++it;
+        
+        std::string output;
+        // Get output if present.
+        if (nodeIsScalarString(it->getKey(), "output")) {
+          if (it->getValue()->getType() != llvm::yaml::Node::NK_Scalar) {
+            error("unexpected 'output' value (expected scalar)");
+            return false;
+          }
+          output = stringFromScalarNode(static_cast<llvm::yaml::ScalarNode*>(it->getValue()));
+          message_out.setOutput(output);
+          ++it;
+        }
+        
+        //get exit status
+        if (!nodeIsScalarString(it->getKey(), "exit-status")) {
+          error("expected key 'exit-status'");
+          return false;
+        }
+        if (it->getValue()->getType() != llvm::yaml::Node::NK_Scalar) {
+          error("unexpected 'exit-status' value (expected scalar)");
+          return false;
+        }
+        
+        int64_t exitStatus;
+        if(!parseInteger(static_cast<llvm::yaml::ScalarNode*>(it->getValue()), &exitStatus)) {
+          return false;
+        }
+        message_out.setExitStatus(exitStatus);
+        ++it;
+        
+        return true;
+        
+      }
+      case Kind::Signalled:
+        // FIXME: parse.
+        return false;
+    }
+    
+    return false;
+  }
+  
+  bool parse(SwiftCompilerOutputMessage& message_out) {
+    llvm::SourceMgr sourceMgr;
+    llvm::yaml::Stream stream(jsonString, sourceMgr);
+    
+    auto it = stream.begin();
+    if (it == stream.end()) {
+      error("missing document in stream");
+      return false;
+    }
+    
+    auto& document = *it;
+    return parseRootNode(document.getRoot(), message_out);
+  }
+  
+};
+
+void SwiftCompilerShellCommand::parseCompilerOutputAndDisplay(std::string jsonString) {
+  SwiftCompilerOutputParserHelper parser(jsonString);
+
+  SwiftCompilerOutputMessage message;
+  
+  // Couldn't parse, nothing to do.
+  if(!parser.parse(message)) {
+    return;
+  }
+
+  if (message.getKind() == SwiftCompilerOutputMessage::Kind::Began && message.getName() == "compile") {
+    llvm::outs() << "Compile " << message.getInputs().front() << "\n";
+    outputmap.insert(std::make_pair(message.getPid(), message));
+  } else if (message.getKind() == SwiftCompilerOutputMessage::Kind::Finished) {
+    if (message.getExitStatus() == 0) {
+      auto it = outputmap.find(message.getPid());
+      if (it != outputmap.end())
+        llvm::outs() << "Compiled " << it->second.getInputs().front()  << "\n";
```

Comment:
I wouldn't bother having two messages for start vs. stop, I think it is too noisy. I have a plan for how we can eventually show the duration of running commands nicely for all tools.